### ISSUE TITLE
Java editor not detecting unused private variable when @Lazy annotated

### ISSUE THREAD

**Issue by r6squeegee** [eclipse-jdt/eclipse.jdt.core#1936]
When unused private variable error/warnings are enabled; annotating a variable with SpringFrameworks @Lazy tag makes the editor think it's used (Where are others like \@Autowired does not)

i.e. 
`
\@Autowired
private FooService fooService;

\@Autowired
\@Lazy 
private BarService barService;
`

if neither fooService nor barService are used at all, only fooService will show the warning/error, barService will not. As soon as you remove the \@Lazy tag it works as expected.

**Comment by stephan-herrmann** [comment]
Should we enumerate every possible annotation out there in internal whitelists or blacklist, so the compiler can respect them for this kind of analysis? That'd be a never-ending story, with lots of discussion whether we are really sure about the semantics of each of those annotations.

Obviously there's an alternative:  let the user decide which annotations to recognize for certain aspects of analysis, by offering some preference options to configure this. Some of the reasons why we were reluctant to do so:
* from a technical p.o.v.: whenever we add more configuration options to the compiler, we need to create two interfaces for this: eclipse preferences and a command line option for ecj. I never felt like the command line is a good vehicle to tell the compiler about (possibly long) lists of well-known types, methods, whatnot and explain how the compiler should respect those during analysis.
* from a social p.o.v.: if users are free to tell the compiler about the semantics of various annotations etc. (to respect them for this analysis, ignore for another aspect), I'm not sure they fully understand the responsibility, but when things go wrong the blame will likely go to ecj, which is not fair :smile:

tl;dr: I'm neither enthusiastic about making this configurable, nor about adding dozens of types (annotations ...) to the internal white lists (except for examples where we have a significant number of user votes).

Unless s.o. has a fresh new idea, I'm leaning towards closing - it may not be perfect, but it works as designed.

**Comment by r6squeegee** [comment]
Maybe I'm just not understanding all the scenarios of annotations properly. I can understand just closing this if it causes people issue or is just too messt. For my own understanding though, could you give me an example as to when a private variable that's unreferenced in the class, would become useful or used due to only an annotation?

i.e. who would encounter a problem if we just ignored the annotation and said - "if a private variable is unreferenced, it's tagged as an unreferenced variable" (regardless of annotations).

**Comment by stephan-herrmann** [comment]
An annotated field could be used at several levels:

1. from Java code within the class
2. from an annotation processor that could use this information for generating some code.
3. from a runtime library using reflection (as is the case, e.g., with injection frameworks).

The compiler can only see level (1). If a user finds a warning about a field being unused, the typical reaction will be to remove the field, which could easily break any application at levels (2) and (3). It might be tempting to say that (2) and (3) will never _read a value from that field_, but the compiler can never be sure about this.

At one end of the spectrum we might even say, _every_ private field can be accessed via reflection, annotated or not, which surely would be spilling the child with the bath. But as soon as a field has an annotation, this signals: there's more to it than meets the eye. That's where we (currently) draw the line.

**Comment by r6squeegee** [comment]
ahhh yes that makes sense, if a annotation post processor write a getter/setter method for that variable, kind of like JPA proxying classes... then the variable will suddenly become important and used.

**Comment by stephan-herrmann** [comment]
Thanks, @r6squeegee  for your feedback.

Closing, since it works as designed.  Still if anyone has a fresh new idea feel free to re-open.
